---
title: "Untitled"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r}
#####Packages#####
library(BoutrosLab.plotting.general)


#####READ TABLE#####

#TCGA prostate cancer RNA-seq data - 2018-05-31_rnaseq_rsem_gene_FPKM.txt
tcga.fpkm.may <- read.delim(
  file = "~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/TCGA_PC/2018-05-31_rnaseq_rsem_gene_FPKM.txt", 
  row.names = 1
  );
#TCGA prostate cancer RNA-seq data - 2018-07-23_rnaseq_rsem_gene_FPKM.txt
tcga.fpkm.july <- read.delim(
  file = "~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/TCGA_PC/2018-07-23_rnaseq_rsem_gene_FPKM.txt", 
  row.names = 1
  );

###Processed CPC-GENE data
#CPC-GENE z-score file calculated using Median
cpcg.zrange.decrease.median <- read.delim(
  file = '~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/cpcg.zrange.decrease.median.txt', 
  row.names = 1
  );
#CPC-GENE z-score file calculated using Trimmed Mean
cpcg.zrange.decrease.trimmean <- read.delim(
  file = '~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/cpcg.zrange.decrease.trimmean.txt', 
  row.names = 1
  );

#Patient part
patient.may <- 1:(ncol(tcga.fpkm.may)-1);
patient.july <- 1:(ncol(tcga.fpkm.july)-1);

#Check the gene order of two files are same
genename.check <- tcga.fpkm.may[,ncol(tcga.fpkm.may)] != tcga.fpkm.july[,ncol(tcga.fpkm.july)];
if (0 != sum(genename.check)) {
  print(tcga.fpkm.may[genename.check, ]);
  }

#Combine two files
tcga.fpkm <- cbind(tcga.fpkm.may[,patient.may], 
                   tcga.fpkm.july);

#Patient part
patient.part <- 1:(ncol(tcga.fpkm)-1);
top.patient <- round(length(patient.part)*0.05, digit = 0);
low.patient <- round(length(patient.part)*0.95, digit = 0);


```

#Calculate z-score usign Median
```{r, fig.width=10, fig.height=7}

#Calculate Median
tcga.median <- function(x) {
   gene <- t(tcga.fpkm[x,patient.part])
   gene.median <- median(gene)
   gene.median
   }
# patient.median <- data.frame();
# for (i in 1:nrow(tcga.fpkm)){
#   patient.median[i,1] <- tcga.median(i)
# };


#Calculate MAD
tcga.mad <- function(x) {
   gene <- t(tcga.fpkm[x,patient.part])
   gene.mad <- mad(gene)
   gene.mad
   }
# patient.mad <- data.frame();
# for (i in 1:nrow(tcga.fpkm)){
#   patient.mad[i,1] <- tcga.mad(i)
# };


#Make Z-score matrix - Using recalculated Median and MAD

zscore.gene.median <- function(x){
   (tcga.fpkm[x,patient.part] - patient.median[x,]) / patient.mad[x,]
   }
# zscore.bymedian <- data.frame();
# for (i in 1 : nrow(tcga.fpkm)) {
#   zscore.bymedian[i,1:length(patient.part)] <- zscore.gene.median(i)
# };
# zscore.bymedian.symbol <- cbind(zscore.bymedian, 
#                                 data.frame(tcga.fpkm$Symbol), 
#                                 patient.median, patient.mad
#                                 );
# colnames(zscore.bymedian.symbol) <- c(colnames(tcga.fpkm), "Median", "mad");
# write.table(
#   zscore.bymedian.symbol, 
#   file = 'tcga.zscore.bymedian.txt', sep = "\t", 
#   col.names = TRUE, 
#   row.names = TRUE, 
#   quote = FALSE
#   );
tcga.zscore.bymedian <- read.delim(
  file = "~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/TCGA_PC/tcga.zscore.bymedian.txt", 
  row.names = 1
  );

#Gene order - have high z-score range: max(z-score)-min(z-score) - all genes : Using Median and MAD
z.max.median <- apply(tcga.zscore.bymedian[patient.part], 1, max);
z.min.median <- apply(tcga.zscore.bymedian[patient.part], 1, min);
z.range.median <- z.max.median - z.min.median;
z.range.median <- data.frame(z.range.median);
tcga.zscore.zrange.median <- cbind(tcga.zscore.bymedian, 
                                   z.range.median);
tcga.zrange.decrease.median <- tcga.zscore.zrange.median[order(tcga.zscore.zrange.median$z.range.median, decreasing = TRUE),];


###Make gene-wise output matrix - maximum FPKM, median z-score, MAD, z-score range, Symbol
#Get maximum FPKM
tcga.max.fpkm <- apply(tcga.fpkm[patient.part], 1, max);
tcga.median.fpkm <- apply(tcga.fpkm[patient.part], 1, median);
#Make gene matrix
tcga.gene.matrix <- cbind(tcga.zscore.bymedian$Median, 
                          tcga.zscore.bymedian$mad, 
                          z.range.median, 
                          tcga.median.fpkm, 
                          tcga.max.fpkm, 
                          tcga.zscore.bymedian$Symbol
                          );
colnames(tcga.gene.matrix) <- c("median", "MAD", "zscore.range", "median.fpkm", "max.fpkm", "Symbol");


#####Check the gene we might lost from median method#####
###Number of genes which have more than half patients with zero FPKM.
nrow(tcga.gene.matrix[is.na(tcga.gene.matrix$zscore.range) & tcga.gene.matrix$max.fpkm > 0,]);
non.median <- tcga.gene.matrix[is.na(tcga.gene.matrix$zscore.range) & tcga.gene.matrix$max.fpkm > 0,]
head(non.median[order(non.median$max.fpkm, decreasing = TRUE),])
high.fpkm.non.median <- as.numeric(tcga.fpkm[rownames(tcga.fpkm) == "ENSG00000164816.7", patient.part]) #gene with highest FPKM 
fpkm.b <- cbind(data.frame(colnames(tcga.fpkm[,patient.part])), high.fpkm.non.median)
colnames(fpkm.b) <- c("patient", "gene")
create.barplot(
      formula = gene ~ patient,
      data = fpkm.b,
      main = "DEFA5",
      main.cex = 1.5,
      ylab.lab = 'FPKM',
      #ylimits = c(0,1),
      #yat = seq(0,1,0.5),
      yaxis.cex = 1,
      ylab.cex = 1,
      xaxis.tck = 0.2,
      #xaxis.lab = NA,
      xlab.lab = "Patient",
      xaxis.rot = 90,
      xaxis.cex = 0.5,
      xlab.cex = 1,
      yaxis.tck = 1,
      yaxis.fontface = 'bold',
      #ylab.axis.padding	= 3,
      sample.order = 'decreasing'
      )
high.fpkm.non.median <- as.numeric(tcga.fpkm[rownames(tcga.fpkm) == "ENSG00000168530.15", patient.part]) #gene with highest FPKM 
fpkm.b <- cbind(data.frame(colnames(tcga.fpkm[,patient.part])), high.fpkm.non.median)
colnames(fpkm.b) <- c("patient", "gene")
create.barplot(
      formula = gene ~ patient,
      data = fpkm.b,
      main = "MYL1",
      main.cex = 1.5,
      ylab.lab = 'FPKM',
      #ylimits = c(0,1),
      #yat = seq(0,1,0.5),
      yaxis.cex = 1,
      ylab.cex = 1,
      xaxis.tck = 0.2,
      #xaxis.lab = NA,
      xlab.lab = "Patient",
      xaxis.rot = 90,
      xaxis.cex = 0.5,
      xlab.cex = 1,
      yaxis.tck = 1,
      yaxis.fontface = 'bold',
      #ylab.axis.padding	= 3,
      sample.order = 'decreasing'
      )


###Make patient-seq output matrix 
#- number of outliers from median (high abundance), number of outliers from median (low abundance), 
thr <- 5; #Z-score threshold
# thre.matrix <- data.frame();
# for (y in 1:nrow(tcga.zscore.bymedian)) {
#   for (i in 1: length(patient.part)) {
#     if (is.na(tcga.zscore.bymedian[y, i])) {
#       thre.matrix[y,i] <- 0
#     } else if (tcga.zscore.bymedian[y, i] > thr) {
#       thre.matrix[y,i] <- 1
#     } else if (tcga.zscore.bymedian[y, i] < -thr) {
#       thre.matrix[y,i] <- -1
#     } else {
#       thre.matrix[y,i] <- 0
#     }
#   }  
# };
# write.table(thre.matrix, 
#             file = 'genewise.outliernumber.bymedian.five.txt',
#             sep = "\t", 
#             col.names = TRUE, 
#             row.names = TRUE, 
#             quote = FALSE
#             );
thre.matrix <- read.delim(
  file = "~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/TCGA_PC/genewise.outliernumber.bymedian.five.txt", 
  row.names = 1
  );
#Calculate the number of outliers for each patients
high.number <- data.frame();
for(i in 1:length(patient.part)){
  high.number[i, 1] <- length(thre.matrix[thre.matrix[,i] == 1 ,i]) 
};
low.number <- data.frame();
for(i in 1:length(patient.part)){
  low.number[i, 1] <- length(thre.matrix[thre.matrix[,i] == -1 ,i]) 
};
#Whole outlier number of each patient
number.outlier <- high.number + low.number;
patient.outlier <- cbind(high.number, 
                         low.number, 
                         number.outlier);
rownames(patient.outlier) <- colnames(tcga.fpkm[,patient.part]);
colnames(patient.outlier) <- c("high", "low", "sum");


#Check maximum value of fpkm > 5 (z-score)
# thr <- 5
# thr.fpkm <- data.frame()
# for (y in 1:100) {
#   for (i in 1: length(patient.part)) {
#     if (is.na(tcga.zscore.bymedian[y, i])) {
#       thr.fpkm[y,i] <- 0
#     } else if (tcga.zscore.bymedian[y, i] > thr) {
#       thr.fpkm[y,i] <- tcga.fpkm[y,i]
#     } else {
#       thr.fpkm[y,i] <- 0
#     }
#   }
# };
# apply(thr.fpkm, 1, max);




```

#Calculate using Median - Make Heatmap
```{r, fig.width=11, fig.height=10}
#Heatmap - for 200 genes
zscore.heat <- create.heatmap(
   x = tcga.zrange.decrease.median[1:200,patient.part],
   clustering.method = 'ward.D2',
   cluster.dimensions = 'both',
   yaxis.lab = NA,
   yaxis.cex = 0.3,
   xaxis.lab = tcga.zrange.decrease.median$Symbol,
   xaxis.cex = 0.3,
   main = "Z-score by Median - Top 200 genes (TCGA-PC)",
   main.cex = 1.5,
   grid.col = FALSE,
   print.colour.key = TRUE,
   yaxis.tck = 0.1,
   xaxis.tck = 0.1,
   colour.scheme = c('red4', 'white', 'dodgerblue4'),
   colour.centering.value = 0,
   at = seq(-5,5,0.001),
   colourkey.cex = 1,
   resolution = 300,
   rows.distance.method = 'euclidean',
   cols.distance.method = 'euclidean'
   );
zscore.heat;


#Make a heatmap in FPKM space
tcga.fpkm.median <- tcga.fpkm[rownames(tcga.fpkm) %in% rownames(tcga.zrange.decrease.median)[1:200],];

tcga.fpkm.median.heat <- create.heatmap(
  x = tcga.fpkm.median[ ,patient.part],
  clustering.method = 'ward.D2',
  #file = generate.filename('', '','png'),
	cluster.dimensions = 'both',
  yaxis.lab = NA,
  yaxis.cex = 0.3,
  xaxis.lab = tcga.fpkm.median$Symbol,
  xaxis.cex = 0.3,
  main = "Top 200 genes by Median from TCGA in FPKM space",
  main.cex = 1.5,
  grid.col = FALSE,
  print.colour.key = TRUE,
  yaxis.tck = 0.1,
  xaxis.tck = 0.1,
  colour.scheme = c('white', 'dodgerblue4'),
  #colourkey.labels.at = seq(-5,5,1),
  at = seq(0,10,0.001),
  colourkey.cex = 1,
  resolution = 300,
  rows.distance.method = 'euclidean',
  cols.distance.method = 'euclidean'
  );
tcga.fpkm.median.heat;
```


#Sort by position - Top 200 genes by median
```{r, fig.width=11, fig.height=10}

library(biomaRt);

#Get gene position
gene.list <- rownames(tcga.zrange.decrease.median[1:200,]);
gene.list.sub <- substr(gene.list, 1, 15);
ensembl <- useEnsembl(biomart = "ensembl", 
                     dataset = "hsapiens_gene_ensembl", 
                     mirror = "uswest");
ensembl <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl);
gene.position <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'chromosome_name',
                                      'start_position', 'end_position', 'band'),
                      filters = 'ensembl_gene_id', 
                      values = gene.list.sub, 
                      mart = ensembl);
  
#Some genes don't match with any position
"%ni%" <- Negate("%in%");
subset(gene.list.sub, gene.list.sub %ni% gene.position$ensembl_gene_id);

#Get the gene, position matrix
tcga.gene.biomart <- subset(tcga.zrange.decrease.median, gene.list.sub %in% gene.position$ensembl_gene_id);
gene.position.order <- gene.position[order(as.numeric(gene.position$chromosome_name), 
                                           as.numeric(gene.position$start_position), 
                                           decreasing = FALSE),];
tcga.gene.biomart.fpkm <- tcga.gene.biomart[match(gene.position.order$ensembl_gene_id, 
                                                  substr(rownames(tcga.gene.biomart), 1, 15)),];
tcga.chr.ord <- cbind(tcga.gene.biomart.fpkm, 
                      gene.position.order[,3:5]);


#Make chromosme covariates : convert "Y"/"X" to numeric value
tcga.chr.ord.numeric <- tcga.chr.ord
for (i in 1 : nrow(tcga.chr.ord.numeric)){
  if(tcga.chr.ord.numeric$chromosome_name[i] == "Y") {
    tcga.chr.ord.numeric$chromosome_name[i] <- "23"
  }
  if(tcga.chr.ord.numeric$chromosome_name[i] == "X") {
    tcga.chr.ord.numeric$chromosome_name[i] <- "24"
  }
};
tcga.chr.ord.numeric.reorder <- tcga.chr.ord.numeric[order(as.numeric(tcga.chr.ord.numeric$chromosome_name), 
                                                           as.numeric(tcga.chr.ord.numeric$start_position), 
                                                           decreasing = FALSE),];
gene.chr <- tcga.chr.ord.numeric.reorder$chromosome_name;
chr <- unique(gene.chr);
cov.colours <- c('darkslategray3','indianred2','darkseagreen1','gold2','honeydew2',
                 'lemonchiffon3','turquoise','khaki','hotpink','darkslateblue', 
                 'indianred4', 'lavender', 'lightblue', 'dodgerblue4', 'darkseagreen3', 
                 'chartreuse3', 'orange', 'lightcoral', 'lightblue4', 'mediumpurple1', 
                 'blueviolet', 'royalblue2', 'ivory2', 'darkslategrey');

sample.col <- cov.colours[as.numeric(gene.chr)];

chr.covariate <- list(
                  rect = list(
                		col = 'transparent',
                		fill = sample.col,
                		lwd = 1.5
                		)
                	);
covariates.legend <- list(
                    	legend = list(
                    		colours = cov.colours,
                    		labels = as.character(c(1:22, "Y", "X")),
                    		title = expression(bold(underline('Chromosome')))
                    		)
                    	);


#Make heatmap by gene position
create.heatmap(
	x = tcga.chr.ord.numeric.reorder[, patient.part],
	clustering.method = 'ward.D2',
  #file = generate.filename('', '','png'),
  cluster.dimensions = 'row',
	main = "Chromosome position - Top 200 genes of TCGA-PC by median",
	main.cex = 1.5,
	covariates.top = chr.covariate,
	covariate.legend = covariates.legend,
	legend.cex = 1,
	legend.title.cex = 1,
	colourkey.cex = 1,
	scale.data = FALSE,
	colour.scheme = c('red4', 'white', 'dodgerblue4'),
  colour.centering.value = 0,
  at = seq(-5,5,0.001),
  yaxis.lab = NA,
  yaxis.cex = 0.3,
  xaxis.lab = tcga.chr.ord.numeric.reorder$Symbol,
  xaxis.cex = 0.3,
	rows.distance.method = 'euclidean',
	resolution = 400
	);

```


#Sort by position - each chromosome by median
```{r, fig.width=11, fig.height=10}

#Get gene position
gene.list.all <- rownames(tcga.zrange.decrease.median);
gene.list.all.sub <- substr(gene.list.all, 1, 15);
ensembl.all <- useEnsembl(biomart = "ensembl", 
                          dataset = "hsapiens_gene_ensembl", 
                          mirror = "uswest");
ensembl.all <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl)
gene.position.all <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'chromosome_name',
                                          'start_position', 'end_position', 'band'),
                           filters = 'ensembl_gene_id', 
                           values = gene.list.all.sub, 
                           mart = ensembl);

#Some genes don't match with any position
"%ni%" <- Negate("%in%");
length(subset(gene.list.all.sub, gene.list.all.sub %ni% gene.position.all$ensembl_gene_id));

#Get the gene, position matrix
tcga.gene.biomart.all <- subset(tcga.zrange.decrease.median, gene.list.all.sub %in% gene.position.all$ensembl_gene_id);
gene.position.order.all <- gene.position.all[order(as.numeric(gene.position.all$chromosome_name),
                                                   as.numeric(gene.position.all$start_position), 
                                                   decreasing = FALSE),];
tcga.gene.biomart.fpkm.all <- tcga.gene.biomart.all[match(gene.position.order.all$ensembl_gene_id, 
                                                          substr(rownames(tcga.gene.biomart.all), 1, 15)),];
tcga.chr.ord.all <- cbind(tcga.gene.biomart.fpkm.all, 
                          gene.position.order.all[,3:5]
                          );
tcga.chr.ord.na <- na.omit(tcga.chr.ord.all);
chr.name <- unique(tcga.chr.ord.na$chromosome_name);

#Assign number for "X", "Y", "MT"
tcga.chr.ord.na.numeric <- tcga.chr.ord.na;
for (i in 1 : nrow(tcga.chr.ord.na.numeric)){
  if(tcga.chr.ord.na.numeric$chromosome_name[i] == "MT") {
    tcga.chr.ord.na.numeric$chromosome_name[i] <- "23"
  }
  if(tcga.chr.ord.na.numeric$chromosome_name[i] == "X") {
    tcga.chr.ord.na.numeric$chromosome_name[i] <- "24"
  }
  if(tcga.chr.ord.na.numeric$chromosome_name[i] == "Y") {
    tcga.chr.ord.na.numeric$chromosome_name[i] <- "25"
  }
};


#Make heatmap for each gene with its chromosomal position
for (i in 1:25){
  heat <- create.heatmap(
  	x = tcga.chr.ord.na.numeric[tcga.chr.ord.na.numeric$chromosome_name == i, patient.part],
  	clustering.method = 'ward.D2',
  	#file = generate.filename('', '','png'),
    cluster.dimensions = 'row',
  	main = paste("Chr", chr.name[i], "- z-score by median from TCGA-PC"),
  	main.cex = 1.5,
  	colourkey.cex = 1,
  	scale.data = FALSE,
  	colour.scheme = c('red4', 'white', 'dodgerblue4'),
    colour.centering.value = 0,
    at = seq(-5,5,0.001),
    yaxis.lab = NA,
    yaxis.cex = 0.3,
  	rows.distance.method = 'euclidean',
  	resolution = 500
  	)
  print(heat)
}


```


#Make Venn Diagram with Top 200 genes by Median from CPC-GENE and TCGA-PC
```{r, fig.width=7, fig.height=7}
#Get top 200 gene list from each dataset
median.symbol.cpcg <- rownames(cpcg.zrange.decrease.median)[1:200];
median.symbol.tcga <- rownames(tcga.zrange.decrease.median)[1:200];


### Venn Diagram
library(VennDiagram);
library(futile.logger);
library(grDevices);

#Top 200 genes from CPC-GENE and TCGA-PC by median
venn <- venn.diagram(
  list(
    CPCG = median.symbol.cpcg, 
    TCGA = median.symbol.tcga
    ),
  main = "Top 200 genes by Median",
  main.fontface = "bold",
  main.fontfamily = "sans",
  main.cex = 2,
  fill = c("firebrick4", "deepskyblue4"), 
  alpha = c(0.35, 0.35),
  cat.fontface = "bold",
  cat.default.pos = "outer",
  #cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055),
  cat.fontfamily = "sans",
  lty = 1,
  lwd = 0.5,
  fontfamily = "sans",
  euler.d = TRUE,
  scaled = TRUE,
  filename = NULL
  );
grid.draw(venn);

```

#Correlation with CPC-GENE data (both calculated using Median)
```{r, fig.width=7, fig.height=7}

#Compare z-score range of CPC-GENE and TCGA-PC data 
# - Using TOP 200 genes from CPC-GENE
subset.median.tcga <- subset(tcga.zrange.decrease.median, 
                             rownames(tcga.zrange.decrease.median) 
                             %in% rownames(cpcg.zrange.decrease.median)[1:200]
                             );
subset.median.tcga.order <- subset.median.tcga[match(rownames(cpcg.zrange.decrease.median)[1:200], 
                                                     rownames(subset.median.tcga)),];
tcga.median.range <- data.frame(subset.median.tcga.order$z.range.median);
cpcg.median.range <- data.frame(cpcg.zrange.decrease.median$z.range.median[1:200]);
subset.zrange.median <- cbind(tcga.median.range, 
                              cpcg.median.range);
rownames(subset.zrange.median) <- rownames(subset.median.tcga.order);
colnames(subset.zrange.median) <- c("TCGA", "CPCGENE");
subset.zrange.median.na <- na.omit(subset.zrange.median); #Some genes have NA
subset.zrange.median.na.log <- log2(subset.zrange.median.na); #Use log2


#Make scatterplot - Top 200 genes
create.scatterplot(
    formula = CPCGENE ~ TCGA,
    data = subset.zrange.median.na.log,
    main = 'Correlation of z-score range by Median - TOP 200 genes from CPC-gene',
    main.cex = 1,
    xlab.label = expression("TCGA-PC log"[2]*"(z-score range)"), 
    ylab.label = expression("CPC-GENE log"[2]*"(z-score range)"),
    xlab.cex = 0.7,
    ylab.cex = 0.7,
    xaxis.cex = 0.5,
    yaxis.cex = 0.5,
    xaxis.fontface = 1,
    yaxis.fontface = 1,
    xlimits = c(0, 20),
    ylimits = c(0, 20),
    xat = seq(0, 20, 5),
    yat = seq(0, 20, 5),
    type = c("p", "r", "g"),
    lwd = 1.5,
    resolution = 300);


#Compare z-score range of CPC-GENE and TCGA-PC data 
# - Using all genes
subset.median.tcga.order.all <- tcga.zrange.decrease.median[match(rownames(cpcg.zrange.decrease.median),
                                                                  rownames(tcga.zrange.decrease.median)),];
tcga.median.range.all <- data.frame(subset.median.tcga.order.all$z.range.median);
cpcg.median.range.all <- data.frame(cpcg.zrange.decrease.median$z.range.median);
subset.zrange.median.all <- cbind(tcga.median.range.all, cpcg.median.range.all);
rownames(subset.zrange.median.all) <- rownames(subset.median.tcga.order.all);
colnames(subset.zrange.median.all) <- c("TCGA", "CPCGENE");
subset.zrange.median.all.na <- na.omit(subset.zrange.median.all); #Some genes have NA
subset.zrange.median.all.na.log <- log2(subset.zrange.median.all.na); #Use log2

#Make scatterplot - all genes
create.scatterplot(
    formula = CPCGENE ~ TCGA,
    data = subset.zrange.median.all.na.log,
    main = 'Correlation of z-score range by Median - All genes',
    main.cex = 1,
    xlab.label = expression("TCGA-PC log"[2]*"(z-score range)"), 
    ylab.label = expression("CPC-GENE log"[2]*"(z-score range)"),
    xlab.cex = 0.7,
    ylab.cex = 0.7,
    xaxis.cex = 0.5,
    yaxis.cex = 0.5,
    xaxis.fontface = 1,
    yaxis.fontface = 1,
    xlimits = c(0, 16.5),
    ylimits = c(0, 16.5),
    xat = seq(0, 16, 4),
    yat = seq(0, 16, 4),
    type = c("p", "r", "g"),
    cex = 0.5,
    lwd = 1.5,
    resolution = 300);

```


#Calculate z-score usign Trimmed Mean
```{r}

#Calculate mean - without the 5% highest and 5% lowest values
tcga.trimmean <- function(x) {
   gene <- t(tcga.fpkm[x,patient.part])
   gene.mean <- mean(gene, trim = 0.05)
   gene.mean
   }

# patient.trimmean <- data.frame();
# for (i in 1:nrow(tcga.fpkm)){
#   patient.trimmean[i,1] <- tcga.trimmean(i)
# };


#Calculate SD - without the 5% highest and 5% lowest values
tcga.tramsd <- function(x) {
   gene <- t(tcga.fpkm[x,patient.part])
   gene.order <- gene[order(gene, decreasing = TRUE),]
   gene.order.sd <- gene.order.SD <- sd(gene.order[(top.patient+1):(low.patient)])
   gene.order.sd
   }
# patient.trimsd <- data.frame();
# for (i in 1:nrow(tcga.fpkm)){
#   patient.trimsd[i,1] <- tcga.tramsd(i)
# };


#Make Z-score matrix - Using recalculated mean and SD : without the 5% highest and 5% lowest values
# zscore.gene.trimmean <- function(x){
#    (tcga.fpkm[x,patient.part] - patient.trimmean[x,]) / patient.trimsd[x,]
#    }
# zscore.bytrimmean <- data.frame();
# for (i in 1 : nrow(tcga.fpkm)) {
#  zscore.bytrimmean[i,1:length(patient.part)] <- zscore.gene.trimmean(i)
# };
# zscore.bytrimmean.symbol <- cbind(zscore.bytrimmean, 
#                                   data.frame(tcga.fpkm$Symbol), 
#                                   patient.trimmean, patient.trimsd);
# colnames(zscore.bytrimmean.symbol) <- c(colnames(tcga.fpkm), "Trimmean", "Trimsd");
# write.table(zscore.bytrimmean.symbol, 
#             file = 'tcga.zscore.bytrimmean.txt', 
#             sep = "\t", 
#             col.names = TRUE, 
#             row.names = TRUE, 
#             quote = FALSE
#             );
tcga.zscore.bytrimmean <- read.delim(
  file = "~/Documents/1.Project/1.CPCGENE_RNA-seq/Rawdata/TCGA_PC/tcga.zscore.bytrimmean.txt", 
  row.names = 1
  );

#Gene order - have high z-score range: max(z-score)-min(z-score) - all genes
z.max.trimmean <- apply(tcga.zscore.bytrimmean[patient.part], 1, max);
z.min.trimmean <- apply(tcga.zscore.bytrimmean[patient.part], 1, min);
z.range.trimmean <- z.max.trimmean - z.min.trimmean;
z.range.trimmean <- data.frame(z.range.trimmean);
tcga.zscore.zrange.trimmean <- cbind(tcga.zscore.bytrimmean, 
                                     z.range.trimmean);
tcga.zrange.decrease.trimmean <- tcga.zscore.zrange.trimmean[order(tcga.zscore.zrange.trimmean$z.range.trimmean, decreasing = TRUE),];
tcga.zrange.decrease.trimmean.inf <- tcga.zrange.decrease.trimmean[is.finite(tcga.zrange.decrease.trimmean$z.range.trimmean),]; #Remove infinite number

```


#Calculate using Trimmed Mean - Make Heatmap
```{r, fig.width=11, fig.height=10}

#Heatmap - for 200 genes
zscore.heat <- create.heatmap(
   x = tcga.zrange.decrease.trimmean.inf[1:200,patient.part],
   clustering.method = 'ward.D2',
   cluster.dimensions = 'both',
   yaxis.lab = NA,
   yaxis.cex = 0.3,
   xaxis.lab = tcga.zrange.decrease.trimmean.inf$Symbol,
   xaxis.cex = 0.3,
   main = "Z-score by Trimmed Mean - Top 200 genes (TCGA-PC)",
   main.cex = 1.5,
   grid.col = FALSE,
   print.colour.key = TRUE,
   yaxis.tck = 0.1,
   xaxis.tck = 0.1,
   colour.scheme = c('red4', 'white', 'dodgerblue4'),
   colour.centering.value = 0,
   at = seq(-5,5,0.001),
   colourkey.cex = 1,
   resolution = 300,
   rows.distance.method = 'euclidean',
   cols.distance.method = 'euclidean'
   );
zscore.heat;

```


#Sort by position - Top 200 genes by Trimmed Mean
```{r, fig.width=11, fig.height=10}

#Get gene position
gene.list <- rownames(tcga.zrange.decrease.trimmean.inf[1:200,]);
gene.list.sub <- substr(gene.list, 1, 15);
ensembl <- useEnsembl(biomart = "ensembl", 
                      dataset = "hsapiens_gene_ensembl", 
                      mirror = "uswest");
ensembl <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl);
gene.position <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'chromosome_name',
                                      'start_position', 'end_position', 'band'),
                       filters = 'ensembl_gene_id', 
                       values = gene.list.sub, 
                       mart = ensembl);
 
#Some genes don't match with any position
"%ni%" <- Negate("%in%");
subset(gene.list.sub, gene.list.sub %ni% gene.position$ensembl_gene_id);

#Get the gene, position matrix
tcga.gene.biomart <- subset(tcga.zrange.decrease.trimmean.inf, gene.list.sub %in% gene.position$ensembl_gene_id);
gene.position.order <- gene.position[order(as.numeric(gene.position$chromosome_name), 
                                           as.numeric(gene.position$start_position), 
                                           decreasing = FALSE),];
tcga.gene.biomart.fpkm <- tcga.gene.biomart[match(gene.position.order$ensembl_gene_id, 
                                                  substr(rownames(tcga.gene.biomart), 1, 15)),];
tcga.chr.ord.trimmean <- cbind(tcga.gene.biomart.fpkm, gene.position.order[,3:5]) ;


#Make chromosme covariates : convert "Y"/"X" to numeric value
tcga.chr.ord.trimmean.numeric <- tcga.chr.ord.trimmean
for (i in 1 : nrow(tcga.chr.ord.trimmean.numeric)){
  if(tcga.chr.ord.trimmean.numeric$chromosome_name[i] == "Y") {
    tcga.chr.ord.trimmean.numeric$chromosome_name[i] <- "23"
  }
  if(tcga.chr.ord.trimmean.numeric$chromosome_name[i] == "X") {
    tcga.chr.ord.trimmean.numeric$chromosome_name[i] <- "24"
  }
};
tcga.chr.ord.numeric.trimmean.reorder <- tcga.chr.ord.trimmean.numeric[order(as.numeric(tcga.chr.ord.trimmean.numeric$chromosome_name), 
                                                           as.numeric(tcga.chr.ord.trimmean.numeric$start_position), 
                                                           decreasing = FALSE),];
gene.chr.trimmean <- tcga.chr.ord.numeric.trimmean.reorder$chromosome_name;
chr.trimmean <- unique(gene.chr.trimmean);
cov.colours <- c('darkslategray3','indianred2','darkseagreen1','gold2','honeydew2',
                 'lemonchiffon3','turquoise','khaki','hotpink','darkslateblue', 
                 'indianred4', 'lavender', 'lightblue', 'dodgerblue4', 'darkseagreen3', 
                 'chartreuse3', 'orange', 'lightcoral', 'lightblue4', 'mediumpurple1', 
                 'blueviolet', 'royalblue2', 'ivory2', 'darkslategrey');

sample.col.trimmean <- cov.colours[as.numeric(gene.chr.trimmean)];
chr.covariate <- list(
                  rect = list(
                		col = 'transparent',
                		fill = sample.col.trimmean,
                		lwd = 1.5
                		)
                	);
covariates.legend <- list(
                    	legend = list(
                    		colours = cov.colours,
                    		labels = as.character(c(1:22, "Y", "X")),
                    		title = expression(bold(underline('Chromosome')))
                    		)
                    	);


#Make heatmap by gene position
create.heatmap(
	x = tcga.chr.ord.numeric.trimmean.reorder[, patient.part],
	clustering.method = 'ward.D2',
  #file = generate.filename('', '','png'),
  cluster.dimensions = 'row',
	main = "Chromosome position - Top 200 genes of TCGA-PC by Trimmed Mean",
	main.cex = 1.5,
	covariates.top = chr.covariate,
	covariate.legend = covariates.legend,
	legend.cex = 1,
	legend.title.cex = 1,
	#same.as.matrix = TRUE,
	colourkey.cex = 1,
	scale.data = FALSE,
	colour.scheme = c('red4', 'white', 'dodgerblue4'),
  colour.centering.value = 0,
  at = seq(-5,5,0.001),
  yaxis.lab = NA,
  yaxis.cex = 0.5,
  xaxis.lab = tcga.chr.ord.numeric.trimmean.reorder$Symbol,
	rows.distance.method = 'euclidean',
  xaxis.cex = 0.3,
	resolution = 400
	);

```


#Sort by position - each chromosome by Trimmed Mean
```{r, fig.width=11, fig.height=10}

#Get gene position
gene.list.all <- rownames(tcga.zrange.decrease.trimmean.inf);
gene.list.all.sub <- substr(gene.list.all, 1, 15);
ensembl.all <- useEnsembl(biomart = "ensembl", 
                          dataset = "hsapiens_gene_ensembl", 
                          mirror = "uswest");
ensembl.all <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl)
gene.position.all <- getBM(attributes = c('ensembl_gene_id', 'hgnc_symbol', 'chromosome_name',
                                          'start_position', 'end_position', 'band'),
                           filters = 'ensembl_gene_id', 
                           values = gene.list.all.sub, 
                           mart = ensembl);

#Some genes don't match with any position
"%ni%" <- Negate("%in%");
length(subset(gene.list.all.sub, gene.list.all.sub %ni% gene.position.all$ensembl_gene_id));

#Get the gene, position matrix
tcga.gene.biomart.trimmean.all <- subset(tcga.zrange.decrease.trimmean.inf, gene.list.all.sub %in% gene.position.all$ensembl_gene_id);
gene.position.order.trimmean.all <- gene.position.all[order(as.numeric(gene.position.all$chromosome_name),
                                                   as.numeric(gene.position.all$start_position), 
                                                   decreasing = FALSE),];
tcga.gene.biomart.fpkm.trimmean.all <- tcga.gene.biomart.trimmean.all[match(gene.position.order.trimmean.all$ensembl_gene_id, 
                                                          substr(rownames(tcga.gene.biomart.trimmean.all), 1, 15)),];
tcga.chr.ord.trimmean.all <- cbind(tcga.gene.biomart.fpkm.trimmean.all, 
                          gene.position.order.trimmean.all[,3:5]
                          );
tcga.chr.ord.trimmean.na <- na.omit(tcga.chr.ord.trimmean.all);
chr.name.trimmean <- unique(tcga.chr.ord.trimmean.na$chromosome_name);

#Assign number for "X", "Y", "MT"
tcga.chr.ord.trimmean.na.numeric <- tcga.chr.ord.trimmean.na;
for (i in 1 : nrow(tcga.chr.ord.trimmean.na.numeric)){
  if(tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] == "MT") {
    tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] <- "23"
  }
  if(tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] == "X") {
    tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] <- "24"
  }
  if(tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] == "Y") {
    tcga.chr.ord.trimmean.na.numeric$chromosome_name[i] <- "25"
  }
};

#Make heatmap for each gene with its chromosomal position
for (i in 1:25){
  heat <- create.heatmap(
  	x = tcga.chr.ord.trimmean.na.numeric[tcga.chr.ord.trimmean.na.numeric$chromosome_name == i, patient.part],
  	clustering.method = 'ward.D2',
  	#file = generate.filename('', '','png'),
    cluster.dimensions = 'row',
  	main = paste("Chr", chr.name.trimmean[i], "- z-score by Trimmed Mean from TCGA-PC"),
  	main.cex = 1.5,
  	colourkey.cex = 1,
  	scale.data = FALSE,
  	colour.scheme = c('red4', 'white', 'dodgerblue4'),
    colour.centering.value = 0,
    at = seq(-5,5,0.001),
    yaxis.lab = NA,
    yaxis.cex = 0.3,
  	rows.distance.method = 'euclidean',
  	resolution = 500
  	)
  print(heat)
}


```



#Make Venn Diagram with Top 200 genes by Trimmed Mean from CPC-GENE and TCGA-PC
```{r, fig.width=7, fig.height=7}
#Get top 200 gene list from each data set
trimmean.symbol.cpcg <- rownames(cpcg.zrange.decrease.trimmean)[1:200];
trimmean.symbol.tcga <- rownames(tcga.zrange.decrease.trimmean.inf)[1:200];


### Venn Diagram
library(VennDiagram);
library(futile.logger);
library(grDevices);

#Top 200 genes from CPC-GENE and TCGA-PC by Trimmed mean
venn <- venn.diagram(
  list(
    CPCG = trimmean.symbol.cpcg, 
    TCGA = trimmean.symbol.tcga
    ),
  main = "Top 200 genes by Trimmed Mean",
  main.fontface = "bold",
  main.fontfamily = "sans",
  main.cex = 2,
  fill = c("firebrick4", "deepskyblue4"), 
  alpha = c(0.35, 0.35),
  cat.fontface = "bold",
  cat.default.pos = "outer",
  #cat.pos = c(-27, 27, 135),
  cat.dist = c(0.055, 0.055),
  cat.fontfamily = "sans",
  lty = 1,
  lwd = 0.5,
  fontfamily = "sans",
  euler.d = TRUE,
  scaled = TRUE,
  filename = NULL
  );
grid.draw(venn);

```

#Correlation with CPC-GENE data (both calculated using Trimmed Mean)
```{r, fig.width=7, fig.height=7}

#Compare z-score range of CPC-GENE and TCGA-PC data 
# - Using TOP 200 genes from CPC-GENE
subset.trimmean.tcga <- subset(tcga.zrange.decrease.trimmean, 
                               rownames(tcga.zrange.decrease.trimmean) 
                               %in% rownames(cpcg.zrange.decrease.trimmean)[1:200]);
subset.trimmean.tcga.order <- subset.trimmean.tcga[match(rownames(cpcg.zrange.decrease.trimmean)[1:200], 
                                                         rownames(subset.trimmean.tcga)),];
tcga.trimmean.range <- data.frame(subset.trimmean.tcga.order$z.range.trimmean);
cpcg.trimmean.range <- data.frame(cpcg.zrange.decrease.trimmean$z.range.trimmean[1:200]);
subset.zrange.trimmean <- cbind(tcga.trimmean.range, 
                                cpcg.trimmean.range);
rownames(subset.zrange.trimmean) <- rownames(subset.trimmean.tcga.order);
colnames(subset.zrange.trimmean) <- c("TCGA", "CPCGENE");
subset.zrange.trimmean.na <- na.omit(subset.zrange.trimmean); #Some genes have NA
subset.zrange.trimmean.na.log <- log2(subset.zrange.trimmean.na); #Use log2


#Make scatterplot
create.scatterplot(
    formula = CPCGENE ~ TCGA,
    data = subset.zrange.trimmean.na.log,
    main = 'Correlation of z-score range by Trimmed Mean - TOP 200 genes from CPC-gene',
    main.cex = 1,
    xlab.label = expression("TCGA-PC log"[2]*"(z-score range)"), 
    ylab.label = expression("CPC-GENE log"[2]*"(z-score range)"),
    xlab.cex = 0.7,
    ylab.cex = 0.7,
    xaxis.cex = 0.5,
    yaxis.cex = 0.5,
    xaxis.fontface = 1,
    yaxis.fontface = 1,
    xlimits = c(0, 20),
    ylimits = c(0, 20),
    xat = seq(0, 20, 5),
    yat = seq(0, 20, 5),
    type = c("p", "r", "g"),
    lwd = 1.5,
    resolution = 300);


#Compare z-score range of CPC-GENE and TCGA-PC data 
# - Using all genes
subset.trimmean.tcga.order.all <- tcga.zrange.decrease.trimmean[match(rownames(cpcg.zrange.decrease.trimmean),
                                                                      rownames(tcga.zrange.decrease.trimmean)),];
tcga.trimmean.range.all <- data.frame(subset.trimmean.tcga.order.all$z.range.trimmean);
cpcg.trimmean.range.all <- data.frame(cpcg.zrange.decrease.trimmean$z.range.trimmean);
subset.zrange.trimmean.all <- cbind(tcga.trimmean.range.all, 
                                    cpcg.trimmean.range.all);
rownames(subset.zrange.trimmean.all) <- rownames(subset.trimmean.tcga.order.all);
colnames(subset.zrange.trimmean.all) <- c("TCGA", "CPCGENE");
subset.zrange.trimmean.all.na <- na.omit(subset.zrange.trimmean.all); #Some genes have NA
subset.zrange.trimmean.all.na.log <- log2(subset.zrange.trimmean.all.na); #Use log2
subset.zrange.trimmean.all.na.log.inf <- subset.zrange.trimmean.all.na.log[is.finite(subset.zrange.trimmean.all.na.log$TCGA),];

#Make scatterplot - all genes
create.scatterplot(
    formula = CPCGENE ~ TCGA,
    data = subset.zrange.trimmean.all.na.log.inf,
    main = 'Correlation of z-score range by Trimmed Mean - All genes',
    main.cex = 1,
    xlab.label = expression("TCGA-PC log"[2]*"(z-score range)"), 
    ylab.label = expression("CPC-GENE log"[2]*"(z-score range)"),
    xlab.cex = 0.7,
    ylab.cex = 0.7,
    xaxis.cex = 0.5,
    yaxis.cex = 0.5,
    xaxis.fontface = 1,
    yaxis.fontface = 1,
    xlimits = c(0, 18),
    ylimits = c(0, 18),
    xat = seq(0, 16, 4),
    yat = seq(0, 16, 4),
    type = c("p", "r", "g"),
    cex = 0.5,
    lwd = 1.5,
    resolution = 300);

```


